### ISSUE TITLE
tf.sqrt() function returns wrong value on Raspberry Pi (3B+) if backend is using "webgl"

### ISSUE THREAD

**Issue by bertlea** [tensorflow/tfjs#1073]
To get help from the community, check out our [Google group](https://groups.google.com/a/tensorflow.org/forum/#!forum/tfjs).

#### TensorFlow.js version
Latest (0.14.1) and also earlier versions.
#### Browser version
Chromium 60.0 (included in latest Raspbian Stretch)
#### Describe the problem or feature request
When testing an portfolio optimization code, it works on all Intel systems but failed when tried to test on Raspberry Pi 3B+ running latest Raspbian with desktops with Chromium browser.   Tracked down the error is the tf.sqrt() function returns a very small value (very close to zero) regardless the value passed in.  This issue only occurs when the back-end is "webgl", which is the default back-end in that environment.  When force to use "cpu", the issue is gone.

Work-around is available as I found I can use tf.pow(tf.scalar(0.5)) instead.  Which gives a value close to the correct value with slightly precision issue compare to using Intel systems or using "cpu" back-end.

#### Code to reproduce the bug / link to feature request
A simple HTML can test that:

`<!DOCTYPE html>
<html>
<head>
	<script src="https://cdn.jsdelivr.net/npm/@tensorflow/tfjs@0.14.1/dist/tf.min.js"></script>
</head>
<body>
	<script>
		const backend = tf.getBackend();
		console.log('calulation enginee: ' + backend);
		document.write('calulation enginee: ' + backend + '<p>');
		const n = tf.scalar(9);
		const ans = n.sqrt().dataSync()[0];
		console.log('answer of square root of 9 = ' + ans);
		document.write('answer of square root of 9 = ' + ans + '<p>');
	</script>
</body>
</html>`

If the code is not formatted right, you can see how to reproduce it from a post I posted to Raspberry Pi forum:  [https://www.raspberrypi.org/forums/viewtopic.php?f=63&t=230632&p=1412807#p1412807](url)

**Comment by bertlea** [comment]
Sorry for the late reply.
@nkreeger , no I am using the default "Legacy" non-GL driver.
@dsmilkov , I got the screenshots.  Basically it seems only WebGL 1 is supported and your debug link seems returns correct sqrt results.
WebGL1:
![webgl1a](https://user-images.githubusercontent.com/33945286/51977443-7499c280-24c2-11e9-878f-beb27fed3e1a.png)
![webgl1b](https://user-images.githubusercontent.com/33945286/51977450-79f70d00-24c2-11e9-8109-70923421d7d1.png)
![webgl1c](https://user-images.githubusercontent.com/33945286/51977505-a01cad00-24c2-11e9-9118-14aa21044c1d.png)
WebGL2:
![webgl2](https://user-images.githubusercontent.com/33945286/51977550-b3c81380-24c2-11e9-8e59-4b4187e84adf.png)
Debug link:
![debug](https://user-images.githubusercontent.com/33945286/51977567-bcb8e500-24c2-11e9-838a-b2fb40e2b1f2.png)

**Comment by bertlea** [comment]
Just tried to switch to the "experimental" GL driver": OpenGL full KMS and that resolved the issue.  So the bug exists only when using the default Legacy Original non-GL desktop driver.

**Comment by nkreeger** [comment]
@bertlea Yes the legacy driver is software-rendering only. Also note that the experimental driver does not support half-float textures (yet) and most of TF.js will not work w/o the minimum requirement of half-float textures (OES_texture_half_float).

**Comment by bertlea** [comment]
Thank you for confirming the cause.  So when tf.js reports it is using 'webgl', it may not actually using the GPU.  The bug in the driver should be a separate issue.
